src/devices/src/virtio/net/simple_handler.rs

```diff
@@ -0,0 +1,188 @@
+use vm_memory::{Bytes, GuestAddressSpace};
+use vm_virtio::{DescriptorChain, Queue};
+
+use crate::virtio::net::tap::Tap;
+use crate::virtio::net::{RXQ_INDEX, TXQ_INDEX};
+use crate::virtio::SignalUsedQueue;
+
+// According to the standard: "If the VIRTIO_NET_F_GUEST_TSO4, VIRTIO_NET_F_GUEST_TSO6 or
+// VIRTIO_NET_F_GUEST_UFO features are used, the maximum incoming packet will be to 65550
+// bytes long (the maximum size of a TCP or UDP packet, plus the 14 byte ethernet header),
+// otherwise 1514 bytes. The 12-byte struct virtio_net_hdr is prepended to this, making for
+// 65562 or 1526 bytes." For transmission, the standard states "The header and packet are added
+// as one output descriptor to the transmitq, and the device is notified of the new entry".
+// We assume the TX frame will not exceed this size either.
+const MAX_BUFFER_SIZE: usize = 65562;
+
+#[derive(Debug)]
+pub enum Error {
+    GuestMemory(vm_memory::GuestMemoryError),
+    Queue(vm_virtio::Error),
+    Tap(io::Error),
+}
+
+impl From<vm_virtio::Error> for Error {
+    fn from(e: vm_virtio::Error) -> Self {
+        Error::Queue(e)
+    }
+}
+
+// A simple handler implementation for a RX/TX queue pair, which does not make assumptions about
+// the way queue notification is implemented. The backend is not yet generic (we always assume a
+// `Tap` object), but we're looking at improving that going forward.
+// TODO: Find a better name.
+pub struct SimpleHandler<M: GuestAddressSpace, S: SignalUsedQueue> {
+    pub driver_notify: S,
+    pub rxq: Queue<M>,
+    pub rxbuf_current: usize,
+    pub rxbuf: [u8; MAX_BUFFER_SIZE],
+    pub txq: Queue<M>,
+    pub txbuf: [u8; MAX_BUFFER_SIZE],
+    pub tap: Tap,
+}
+
+impl<M: GuestAddressSpace, S: SignalUsedQueue> SimpleHandler<M, S> {
+    pub fn new(driver_notify: S, rxq: Queue<M>, txq: Queue<M>, tap: Tap) -> Self {
+        SimpleHandler {
+            driver_notify,
+            rxq,
+            rxbuf_current: 0,
+            rxbuf: [0u8; MAX_BUFFER_SIZE],
+            txq,
+            txbuf: [0u8; MAX_BUFFER_SIZE],
+            tap,
+        }
+    }
+
+    // Have to see how to approach error handling for the `Queue` implementation in particular,
+    // because many situations are not really recoverable. We should consider reporting them based
+    // on the  metrics/events solution when they appear, and not propagate them further unless
+    // it's really useful/necessary.
+    fn write_frame_to_guest(&mut self) -> result::Result<bool, Error> {
+        let num_bytes = self.rxbuf_current;
+
+        let mut chain = match self.rxq.iter()?.next() {
+            Some(c) => c,
+            _ => return Ok(false),
+        };
+
+        let mut count = 0;
+        let buf = &mut self.rxbuf[..num_bytes];
+
+        while let Some(desc) = chain.next() {
+            let left = buf.len() - count;
+
+            if left == 0 {
+                break;
+            }
+
+            let len = cmp::min(left, desc.len() as usize);
```

Comment:
What happens if `len` field from the descriptor is always 0 and we don't write any data to memory (count = 0 at the end of this loop)? Does it still make sense to add an entry in the used ring when no data was actually transferred?